File: Sources/ExcalidrawMath/Point.swift
import Foundation

/// A 2D point in scene coordinates.
///
/// Mirrors Excalidraw's `LocalPoint`/`GlobalPoint` tuple `[number, number]`
/// (see `packages/math/src/point.ts`). Encodes to/from a two-element JSON
/// array `[x, y]` to match the `.excalidraw` wire format (points, scale,
/// fixedPoint, fixedSegment endpoints are all stored this way).
public struct Point: Equatable, Hashable, Sendable {
    public var x: Double
    public var y: Double

    public init(_ x: Double, _ y: Double) {
        self.x = x
        self.y = y
    }

    /// Euclidean distance to another point.
    public func distance(to other: Point) -> Double {
        (self - other).magnitude
    }

    /// Vector magnitude treating the point as a vector from the origin.
    public var magnitude: Double {
        (x * x + y * y).squareRoot()
    }

    public static func - (lhs: Point, rhs: Point) -> Point {
        Point(lhs.x - rhs.x, lhs.y - rhs.y)
    }

    public static func + (lhs: Point, rhs: Point) -> Point {
        Point(lhs.x + rhs.x, lhs.y + rhs.y)
    }
}

extension Point: Codable {
    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        let x = try container.decode(Double.self)
        let y = try container.decode(Double.self)
        self.init(x, y)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.unkeyedContainer()
        try container.encode(x)
        try container.encode(y)
    }
}

public extension Point {
    /// The origin, `(0, 0)`.
    static let zero = Point(0, 0)

    /// Rotate this point around `center` by `angle` radians (`pointRotateRads`).
    func rotated(around center: Point, by angle: Double) -> Point {
        guard angle != 0 else { return self }
        let dx = x - center.x
        let dy = y - center.y
        return Point(
            dx * cos(angle) - dy * sin(angle) + center.x,
            dx * sin(angle) + dy * cos(angle) + center.y
        )
    }

    /// Translate this point by a vector (`pointTranslate`).
    func translated(by v: Vector) -> Point {
        Point(x + v.u, y + v.v)
    }

    /// The midpoint between this point and `other` (`pointCenter`).
    func midpoint(to other: Point) -> Point {
        Point((x + other.x) / 2, (y + other.y) / 2)
    }

    /// Squared Euclidean distance (`pointDistanceSq`) — avoids a `sqrt`.
    func distanceSquared(to other: Point) -> Double {
        let dx = other.x - x
        let dy = other.y - y
        return dx * dx + dy * dy
    }

    /// Scale this point away from `origin` by `multiplier` (`pointScaleFromOrigin`).
    func scaled(from origin: Point, by multiplier: Double) -> Point {
        origin.translated(by: Vector(from: self, origin: origin).scaled(by: multiplier))
    }

    /// Whether this point lies within the axis-aligned box bounded by `a` and
    /// `b` (`isPointWithinBounds`).
    func isWithin(_ a: Point, _ b: Point) -> Bool {
        x <= Swift.max(a.x, b.x) && x >= Swift.min(a.x, b.x) &&
            y <= Swift.max(a.y, b.y) && y >= Swift.min(a.y, b.y)
    }

    /// Coordinate-wise comparison within `tolerance` (`pointsEqual`).
    func isApproximatelyEqual(to other: Point, tolerance: Double = ExcalidrawMath.precision) -> Bool {
        abs(x - other.x) < tolerance && abs(y - other.y) < tolerance
    }
}

public enum ExcalidrawMath {
    /// Numerical tolerance used throughout geometry, matching upstream
    /// `PRECISION` (`10e-5`, i.e. `1e-4`).
    public static let precision = 10e-5
}

File: web/packages/excalidraw-svelte/src/math/point.ts
import { Vector } from "./vector.js";

/**
 * Numerical tolerance used throughout geometry, matching upstream `PRECISION`
 * (`10e-5`, i.e. `1e-4`). (parity: ExcalidrawMath.precision)
 */
export const PRECISION = 10e-5;

/**
 * A 2D point in scene coordinates. Wire form is the JSON array `[x, y]`, to
 * match the `.excalidraw` format (points, scale, fixedPoint, fixedSegment).
 * (parity: Sources/ExcalidrawMath/Point.swift)
 */
export class Point {
  constructor(
    public x: number,
    public y: number,
  ) {}

  static readonly zero = new Point(0, 0);

  /** Euclidean distance to another point. */
  distance(other: Point): number {
    return this.sub(other).magnitude;
  }

  /** Magnitude treating the point as a vector from the origin. */
  get magnitude(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  add(other: Point): Point {
    return new Point(this.x + other.x, this.y + other.y);
  }

  sub(other: Point): Point {
    return new Point(this.x - other.x, this.y - other.y);
  }

  /** Rotate around `center` by `angle` radians (`pointRotateRads`). */
  rotated(center: Point, angle: number): Point {
    if (angle === 0) return this;
    const dx = this.x - center.x;
    const dy = this.y - center.y;
    return new Point(
      dx * Math.cos(angle) - dy * Math.sin(angle) + center.x,
      dx * Math.sin(angle) + dy * Math.cos(angle) + center.y,
    );
  }

  /** Translate by a vector (`pointTranslate`). */
  translated(v: Vector): Point {
    return new Point(this.x + v.u, this.y + v.v);
  }

  /** Midpoint between this and `other` (`pointCenter`). */
  midpoint(other: Point): Point {
    return new Point((this.x + other.x) / 2, (this.y + other.y) / 2);
  }

  /** Squared Euclidean distance — avoids a `sqrt` (`pointDistanceSq`). */
  distanceSquared(other: Point): number {
    const dx = other.x - this.x;
    const dy = other.y - this.y;
    return dx * dx + dy * dy;
  }

  /** Scale away from `origin` by `multiplier` (`pointScaleFromOrigin`). */
  scaledFrom(origin: Point, multiplier: number): Point {
    return origin.translated(Vector.fromPoint(this, origin).scaled(multiplier));
  }

  /** Whether this lies within the AABB bounded by `a` and `b`. */
  isWithin(a: Point, b: Point): boolean {
    return (
      this.x <= Math.max(a.x, b.x) &&
      this.x >= Math.min(a.x, b.x) &&
      this.y <= Math.max(a.y, b.y) &&
      this.y >= Math.min(a.y, b.y)
    );
  }

  /** Coordinate-wise comparison within `tolerance` (`pointsEqual`). */
  isApproximatelyEqual(other: Point, tolerance = PRECISION): boolean {
    return Math.abs(this.x - other.x) < tolerance && Math.abs(this.y - other.y) < tolerance;
  }

  /** Exact equality. */
  equals(other: Point): boolean {
    return this.x === other.x && this.y === other.y;
  }

  /** Wire form `[x, y]`. */
  toArray(): [number, number] {
    return [this.x, this.y];
  }

  static fromArray([x, y]: readonly [number, number]): Point {
    return new Point(x, y);
  }
}
